// tree-kit/lib/path.js
/*
	The Cedric's Swiss Knife (CSK) - CSK object tree toolbox

	Copyright (c) 2014, 2015 Cédric Ronvel 
	
	The MIT License (MIT)

	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:

	The above copyright notice and this permission notice shall be included in all
	copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
	SOFTWARE.
*/



var treePath = {} ;
exports.path = treePath ;



treePath.op = function op( type , object , path , value )
{
	var i , parts , last , pointer , key , isArray = false , pathArrayMode = false , isGenericSet ;
	
	if ( typeof path === 'string' )
	{
		// Split the path into parts
		parts = path.match( /([.#\[\]]|[^.#\[\]]+)/g ) ;
		//parts = path.match( /([.#](?!$)|[^.#]+)/g ) ;
	}
	else if ( Array.isArray( path ) )
	{
		parts = path ;
		pathArrayMode = true ;
	}
	else
	{
		throw new TypeError( '[tree.path] .' + type + '(): the path argument should be a string or an array' ) ;
	}
	
	switch ( type )
	{
		case 'get' :
		case 'delete' :
			isGenericSet = false ;
			break ;
		case 'set' :
		case 'define' :
		case 'inc' :
		case 'dec' :
			isGenericSet = true ;
			break ;
		default :
			throw new TypeError( "[tree.path] .op(): wrong type of operation '" + type + "'" ) ;
	}
	
	
	//console.log( parts ) ;
	// The pointer start at the object's root
	pointer = object ;
	
	last = parts.length - 1 ;
	
	for ( i = 0 ; i <= last ; i ++ )
	{
		if ( pathArrayMode )
		{
			if ( key === undefined )
			{
				key = parts[ i ] ;
				continue ;
			}
			
			if ( ! pointer[ key ] || ( typeof pointer[ key ] !== 'object' && typeof pointer[ key ] !== 'function' ) )
			{
				if ( ! isGenericSet ) { return undefined ; }
				pointer[ key ] = {} ;
			}
			
			pointer = pointer[ key ] ;
			key = parts[ i ] ;
			
			continue ;
		}
		else if ( parts[ i ] === '.' )
		{
			isArray = false ;
			
			if ( key === undefined ) { continue ; }
			
			if ( ! pointer[ key ] || ( typeof pointer[ key ] !== 'object' && typeof pointer[ key ] !== 'function' ) )
			{
				if ( ! isGenericSet ) { return undefined ; }
				pointer[ key ] = {} ;
			}
			
			pointer = pointer[ key ] ;
			
			continue ;
		}
		else if ( parts[ i ] === '#' || parts[ i ] === '[' )
		{
			isArray = true ;
			
			if ( key === undefined )
			{
				// The root element cannot be altered, we are in trouble if an array is expected but we have only a regular object.
				if ( ! Array.isArray( pointer ) ) { return undefined ; }
				continue ;
			}
			
			if ( ! pointer[ key ] || ! Array.isArray( pointer[ key ] ) )
			{
				if ( ! isGenericSet ) { return undefined ; }
				pointer[ key ] = [] ;
			}
			
			pointer = pointer[ key ] ;
			
			continue ;
		}
		else if ( parts[ i ] === ']' )
		{
			// Closing bracket: do nothing
			continue ;
		}
		
		if ( ! isArray ) { key = parts[ i ] ; continue ; }
		
		switch ( parts[ i ] )
		{
			case 'length' :
				key = parts[ i ] ;
				break ;
			
			// Pseudo-key
			case 'first' :
				key = 0 ;
				break ;
			case 'last' :
				key = pointer.length - 1 ;
				if ( key < 0 ) { key = 0 ; }
				break ;
			case 'next' :
				if ( ! isGenericSet ) { return undefined ; }
				key = pointer.length ;
				break ;
			case 'insert' :
				if ( ! isGenericSet ) { return undefined ; }
				pointer.unshift( undefined ) ;
				key = 0 ;
				break ;
			
			// default = number
			default:
				// Convert the string key to a numerical index
				key = parseInt( parts[ i ] , 10 ) ;
		}
	}
	
	switch ( type )
	{
		case 'get' :
			return pointer[ key ] ;
		case 'delete' :
			if ( isArray && typeof key === 'number' ) { pointer.splice( key , 1 ) ; }
			else { delete pointer[ key ] ; }
			return ;
		case 'set' :
			pointer[ key ] = value ;
			return pointer[ key ] ;
		case 'define' :
			// define: set only if it doesn't exist
			if ( ! ( key in pointer ) ) { pointer[ key ] = value ; }
			return pointer[ key ] ;
		case 'inc' :
			if ( typeof pointer[ key ] === 'number' ) { pointer[ key ] ++ ; }
			else if ( ! pointer[ key ] || typeof pointer[ key ] !== 'object' ) { pointer[ key ] = 1 ; }
			return pointer[ key ] ;
		case 'dec' :
			if ( typeof pointer[ key ] === 'number' ) { pointer[ key ] -- ; }
			else if ( ! pointer[ key ] || typeof pointer[ key ] !== 'object' ) { pointer[ key ] = -1 ; }
			return pointer[ key ] ;
	}
} ;



// get, set and delete use the same op() function
treePath.get = treePath.op.bind( undefined , 'get' ) ;
treePath.delete = treePath.op.bind( undefined , 'delete' ) ;
treePath.set = treePath.op.bind( undefined , 'set' ) ;
treePath.define = treePath.op.bind( undefined , 'define' ) ;
treePath.inc = treePath.op.bind( undefined , 'inc' ) ;
treePath.dec = treePath.op.bind( undefined , 'dec' ) ;



// Prototype used for object creation, so they can be created with Object.create( tree.path.prototype )
treePath.prototype = {
	get: function( path ) { return treePath.get( this , path ) ; } ,
	"delete": function( path ) { return treePath.delete( this , path ) ; } ,
	set: function( path , value ) { return treePath.set( this , path , value ) ; } ,
	define: function( path , value ) { return treePath.define( this , path , value ) ; } ,
	inc: function( path , value ) { return treePath.inc( this , path , value ) ; } ,
	dec: function( path , value ) { return treePath.dec( this , path , value ) ; }
} ;



// Upgrade an object so it can support get, set and delete at its root
treePath.upgrade = function upgrade( object )
{
	Object.defineProperties( object , {
		get: { value: treePath.op.bind( undefined , 'get' , object ) } ,
		"delete": { value: treePath.op.bind( undefined , 'delete' , object ) } ,
		set: { value: treePath.op.bind( undefined , 'set' , object ) } ,
		define: { value: treePath.op.bind( undefined , 'define' , object ) } ,
		inc: { value: treePath.op.bind( undefined , 'inc' , object ) } ,
		dec: { value: treePath.op.bind( undefined , 'dec' , object ) }
	} ) ;
} ;
